### PR TITLE
Implement variable-sized allocation of ID strings

### PR DESCRIPTION
**Pull request by dmgaldi**
## Overview
* Update file dumper tool to allocate ID string based on max length of entity ID string
* Refactor some code into lib-subsetting

## Testing
* Ran a full dump of files on my Mac and ran regression tests against using dumped files

### CODE REVIEW

File: src/main/java/org/veupathdb/eda/binaryfiles/dumper/IdFilesDumperFactory.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    this.entity = entity;
+    this.parentEntity = parentEntity;
+  }
+
+  /**
+   * Supplies an implementation of an ID FilesDumper based on the number of ancestors of entity member and the
+   * byte count allocated for IDs in ancestry.
+   */
+  public FilesDumper create(Map<String, Integer> bytesReservedForIdByEntityId) {
+    if (entity.getAncestorEntities().size() == 1) {
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/eda/binaryfiles/dumper/IdFilesDumperFactory.java:30]: Could return a switch here: case 0, case 1, default.  Not sure what would be clearer, but nice to not parse if conditions.

File: src/main/java/org/veupathdb/eda/binaryfiles/dumper/MaxIdLengthFinder.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  private static final int INDEX_OF_ID = 0;
+  private int maxLength = 0;
+  private boolean done = false;
+  private boolean firstRow = true;
+
+  @Override
+  public void consumeRow(List<String> record) throws IOException {
+    if (firstRow) { firstRow = false; return; } // skip header
+
+    int numBytes = record.get(INDEX_OF_ID).getBytes(StandardCharsets.UTF_8).length + 4;
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/eda/binaryfiles/dumper/MaxIdLengthFinder.java:22]: Add a comment here about why you add 4 at the end.

File: src/main/java/org/veupathdb/eda/binaryfiles/dumper/StudyDumper.java [hunk 3]
```diff
@@ -42,33 +44,48 @@ public void dumpStudy() {
...
     Entity entity = subTree.getContents();
-    
-    dumpEntity(entity, idsDumperSupplier);
+
+    // Find the maximum length of an ID for this entity to determine space to allocate for each ID.
+    // Add 4 to store the size of the ID as well.
+    int maxIdLength = scanForMaxIdLength(entity);
+
+    // Add the value to the Map so that it is available when dumping subtrees recursively.
+    Map<String, Integer> newMaxIdLengthMap = new HashMap(entityIdToMaxIdLength);
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/eda/binaryfiles/dumper/StudyDumper.java:59]: Why make a copy here?  Cheap, but necessary?  The entities inside will be added top-down so I could see only wanting the actual ancestors of the "current" entity, but then you loop over the ancestors below at 79 (not use e.g. new JSONObject(entityIdToMaxLength)).  Just trying to understand- maybe a comment as to why a copy is preferred?

File: src/main/java/org/veupathdb/eda/binaryfiles/dumper/multi/Main.java [hunk 4]
```diff
@@ -38,17 +41,40 @@ public static void main(String[] args) throws Exception {
...
+    if (!failedStudies.isEmpty()) {
+      final String commaSeparatedStudies = failedStudies.stream()
+          .map(study -> study.studyOverview.getStudyId())
+          .collect(Collectors.joining(","));
+      throw new RuntimeException("The following studies failed to be written: " + commaSeparatedStudies
+          + ". Attaching first failure stack trace: ", failedStudies.stream().findFirst().get().e);
+    }
+  }
+
+  public static class FailedStudy {
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/eda/binaryfiles/dumper/multi/Main.java:71]: I often extend TwoTuple in cases like this to save a few lines.  Members are available via getFirst/getKey and getSecond/getValue.  Not sure if it's a bad habit though. e.g.
```
private static class FailedStudy extends TwoTuple<Exception, StudyOverview> {
    public FailedStudy(Exception first, StudyOverview second) { super(first, second); }
}

File: src/main/java/org/veupathdb/eda/binaryfiles/printer/BinaryFilePrinter.java [hunk 5]
```diff
@@ -86,15 +78,19 @@ private static void printAncestorFile(Path binaryFile, JSONObject metajson) {
...
       } catch (IOException e) {
-      throw new RuntimeException("Failed attempting to read file " + binaryFile.toString(), e);
+      throw new RuntimeException("Failed attempting to read file " + binaryFile, e);
     }
   }    
   
   private static void printIdsMapFile(Path binaryFile, JSONObject metajson) {
+    List<Integer> bytesReservedPerAncestors = new ArrayList<>();
+    int bytesReservedForId = metajson.getInt("bytesReservedForId");
+    JSONArray ancestorBytesReserved = metajson.getJSONArray("bytesReservedPerAncestor");
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/eda/binaryfiles/printer/BinaryFilePrinter.java:88]: These property names should probably be constants.  Missed where they are set but should be used there too.

**Review by ryanrdoherty — APPROVED** [review]

File: pom.xml [hunk 6]
```diff
@@ -31,7 +31,7 @@
     <dependency>
       <groupId>org.veupathdb.eda</groupId>
       <artifactId>lib-eda-subsetting</artifactId>
-      <version>1.8.1</version>
+      <version>3.0.0</version>
```
[the remark this answers is not in the document]
Comment (dmgaldi) [reply pom.xml:34]: We've been pretty aggressive about making backward-incompatible changes. Mostly just class name changes as opposed to anything "major".

File: src/main/java/org/veupathdb/eda/binaryfiles/dumper/IdFilesDumperFactory.java (hunk 1 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/eda/binaryfiles/dumper/IdFilesDumperFactory.java:30]: Sure, I don't feel strongly but that's a bit more terse

File: src/main/java/org/veupathdb/eda/binaryfiles/dumper/MaxIdLengthFinder.java (hunk 2 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/eda/binaryfiles/dumper/MaxIdLengthFinder.java:22]: Added, also changed the 4 to Integer.BYTES for clarity

File: src/main/java/org/veupathdb/eda/binaryfiles/dumper/StudyDumper.java (hunk 3 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/eda/binaryfiles/dumper/StudyDumper.java:59]: Not necessary, I think I just did it out of paranoia about the side effect even though it shouldn't be an issue. I'll avoid the copy to make it less confusing

File: src/main/java/org/veupathdb/eda/binaryfiles/dumper/multi/Main.java (hunk 4 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/eda/binaryfiles/dumper/multi/Main.java:71]: I might slightly prefer to keep as is to make the getters more clear. I'm open to update if you feel strongly about it though

File: src/main/java/org/veupathdb/eda/binaryfiles/printer/BinaryFilePrinter.java (hunk 5 above)
Comment (dmgaldi) [reply src/main/java/org/veupathdb/eda/binaryfiles/printer/BinaryFilePrinter.java:88]: Thanks, I extracted them to constants and missed updating here.

**Pull request merged**